File: Jena/operation/graph/TriangleCounter.py
'''
Created on Jan 21, 2016

@author: Seokyong Hong
'''
from threading import Thread
from operation.GraphOperation import GraphOperation

counter = []

class Worker(Thread):
    def __init__(self, name, query, connection):
        Thread.__init__(self)
        self.name = name
        self.query = query
        self.connection = connection
    
    def run(self):
        global counter
        
        results = self.connection.urika.query(self.name, self.query, None, None, 'json', True)
        counter.append(int(results['results']['bindings'][0]['count']['value']))

class TriangleCounter(GraphOperation):
    COMMAND_TO_INITIALIZE = '''
            CREATE GRAPH <http://workingGraph>;
                            '''
    COMMAND_TO_FINALIZE = '''
            DROP GRAPH <http://workingGraph>;
                          '''
    COMMAND_TO_STORE_INITIAL_GRAPH = '''
            INSERT { GRAPH <http://workingGraph> { ?node <temp:links> ?neighbor }}
            WHERE {
                SELECT ?node ?neighbor
                WHERE {
                    ?node <urn:connectedTo>|^<urn:connectedTo> ?neighbor .
                    FILTER (STR(?node) < STR(?neighbor))
                }
            };
                                     '''
    COMMAND_TO_COUNT_TRIANGLES = '''
            SELECT (COUNT(DISTINCT *) AS ?count)
            WHERE {
                GRAPH <http://workingGraph> {
                     ?x <temp:links> ?y .
                     ?y <temp:links> ?z .
                     ?x <temp:links> ?z
                }
            }
                                 '''
    COMMAND_TO_COUNT_TRIANGLES_DIRECTED_SINGLE = '''
            SELECT (COUNT(*) AS ?count)
                WHERE {
                    {
                        ?x ?a ?y .
                        ?y ?b ?z .
                        ?z ?c ?x .
                        FILTER (STR(?x) < STR(?y)) .
                        FILTER (STR(?y) < STR(?z))
                    }
                    UNION 
                    {
                        ?x ?a ?y .
                        ?y ?b ?z .
                        ?z ?c ?x .
                        FILTER (STR(?y) > STR(?z)) .
                        FILTER (STR(?z) > STR(?x))
                    }  
                    UNION 
                    {
                        ?x ?a ?y .
                        ?y ?b ?z .
                        ?x ?c ?z
                    }
                }
                                                 '''
    COMMANDS_FOR_MULTITHREADS = [
                  '''
            SELECT (COUNT(*) AS ?count)
            WHERE {
                ?x ?a ?y .
                ?y ?b ?z .
                ?z ?c ?x .
                FILTER (STR(?x) < STR(?y)) .
                FILTER (STR(?y) < STR(?z))
            }
                  '''
                  ,
                  '''
            SELECT (COUNT(*) AS ?count)
            WHERE {
                ?x ?a ?y .
                ?y ?b ?z .
                ?z ?c ?x .
                FILTER (STR(?y) > STR(?z)) .
                FILTER (STR(?z) > STR(?x)) 
            }
                  '''
                  ,
                  '''
            SELECT (COUNT(*) AS ?count)
            WHERE {
                ?x ?a ?y .
                ?y ?b ?z .
                ?x ?c ?z
            }
                  '''
    ] 
    
    def __init__(self, directed, multithreaded, connection):
        GraphOperation.__init__(self, 'Triangle Counter', connection)
        self.directed = directed
        self.multithreaded = multithreaded
    
    def initialize(self):
        if self.directed == False:
            self.connection.urika.update(self.name, TriangleCounter.COMMAND_TO_INITIALIZE)
        else:
            if self.multithreaded == True:
                self.workers = []
                for worker in range(0, 3):
                    self.workers.append(Worker('Worker-' + str(worker + 1), TriangleCounter.COMMANDS_FOR_MULTITHREADS[worker], self.connection))
            
    def process(self):
        if self.directed == False:
            self.connection.urika.update(self.name, TriangleCounter.COMMAND_TO_STORE_INITIAL_GRAPH)
            results = self.connection.urika.query(self.name, TriangleCounter.COMMAND_TO_COUNT_TRIANGLES, None, None, 'json', True)
            self.count = results['results']['bindings'][0]['count']['value']
        else:
            if self.multithreaded == False:
                results = self.connection.urika.query(self.name, TriangleCounter.COMMAND_TO_COUNT_TRIANGLES_DIRECTED_SINGLE, None, None, 'json', True)
                self.count = results['results']['bindings'][0]['count']['value']
            else:
                for worker in self.workers:
                    worker.start();
            
                for worker in self.workers:
                    worker.join()
                
                global counter
                self.count = 0
                for count in counter:
                    self.count = self.count + count
       
    def store(self):
        if self.writer is not None:
            self.writer.writeLine('Number of triangles: ' + str(self.count))
    
    def finalize(self):
        if self.directed == False:
            self.connection.urika.update(self.name, TriangleCounter.COMMAND_TO_FINALIZE)

File: Sesame/src/operation/graph/TriangleCounter.java
package operation.graph;

import org.openrdf.model.Literal;
import org.openrdf.query.BindingSet;
import org.openrdf.query.TupleQueryResult;
import org.openrdf.repository.Repository;
import org.openrdf.repository.RepositoryConnection;

import exception.OperationException;
import operation.GraphOperation;

public class TriangleCounter extends GraphOperation {
	private static final String NAME = "Triangle Counting";
	
	private boolean isDirected;
	private boolean isMultithreaded;
	private Thread workers[];
	private int counts[];
	
	private static final String QUERY_INIT_GRAPH = ""
			+ "CREATE GRAPH <http://workingGraph>;";
	
	private static final String QUERY_STORE_INITIAL_GRAPH = ""
			+ "INSERT { GRAPH <http://workingGraph> { ?vertex <temp:links> ?neighbor }}\n"
			+ "WHERE {\n"
			+ "  SELECT ?vertex ?neighbor\n"
			+ "  WHERE {\n"
			+ "    ?vertex <urn:connectedTo>|^<urn:connectedTo> ?neighbor .\n"
			+ "    FILTER (STR(?vertex) < STR(?neighbor))\n"
			+ "  }\n"
			+ "};";
	private static final String QUERY_FIND_TRIANGLES = ""
			+ "SELECT (COUNT(DISTINCT *) AS ?count)\n"
			+ "WHERE {\n"
			+ "  GRAPH <http://workingGraph> {\n"
			+ "    ?x ?p ?y .\n"
			+ "    ?y ?p ?z .\n"
			+ "    ?x ?p ?z\n"
			+ "  }\n"
			+ "}";
	private static final String QUERY_FIND_TRIANGLES_FOR_DIRECTED_GRAPH = ""
			+ "SELECT (COUNT(*) AS ?count)\n"
            + "WHERE {\n"
			+ "  {\n"
            + "    ?x ?a ?y .\n"
			+ "    ?y ?b ?z .\n"
            + "    ?z ?c ?x .\n"
			+ "    FILTER (STR(?x) < STR(?y)) .\n"
            + "    FILTER (STR(?y) < STR(?z))\n"
			+ "  }\n"
            + "  UNION\n"
			+ "  {\n"
            + "    ?x ?a ?y .\n"
			+ "    ?y ?b ?z .\n"
            + "    ?z ?c ?x .\n"
			+ "    FILTER (STR(?y) > STR(?z)) .\n"
            + "    FILTER (STR(?z) > STR(?x))\n"
			+ "  }\n"
            + "  UNION\n"
			+ "  {\n"
            + "    ?x ?a ?y .\n"
			+ "    ?y ?b ?z .\n"
            + "    ?x ?c ?z\n"
			+ "  }\n"
            + "}";
	private static final String[] QUERY_FOR_MULTITHREADS = {
			"SELECT (COUNT(*) AS ?count)\n" +
			"WHERE {\n" +
			"  ?x ?a ?y .\n" +
			"  ?y ?b ?z .\n" +
			"  ?z ?c ?x .\n" +
			"  FILTER (STR(?x) < STR(?y)) .\n" +
			"  FILTER (STR(?y) < STR(?z))\n" +
			"}",
			"SELECT (COUNT(*) AS ?count)\n" +
			"WHERE {\n" +
			"  ?x ?a ?y .\n" +
			"  ?y ?b ?z .\n" +
			"  ?z ?c ?x .\n" +
			"  FILTER (STR(?y) > STR(?z)) .\n" +
			"  FILTER (STR(?z) > STR(?x))\n" +
			"}",
			"SELECT (COUNT(*) AS ?count)\n" +
			"WHERE {\n" +
			"  ?x ?a ?y .\n" +
			"  ?y ?b ?z .\n" +
			"  ?x ?c ?z\n" +
			"}"
	};
	private static final String QUERY_DROP_GRAPH = ""
			+ "DROP GRAPH <http://workingGraph>;";
	
	public TriangleCounter(RepositoryConnection connection, boolean isDirected, boolean isMultithreaded) throws OperationException {
		super(NAME, connection);
		this.isDirected = isDirected;
		this.isMultithreaded = isMultithreaded;
		if(isDirected && isMultithreaded) {
			workers = new Worker[3];
			counts = new int[3];
		}
	}
	
	@Override
	protected void init() {
		if(!this.isDirected)
			connection.prepareUpdate(QUERY_INIT_GRAPH).execute();
		else {
			if(this.isDirected && this.isMultithreaded) {
				Repository repository = connection.getRepository();
				for(int index = 0;index < workers.length;index++) {
					workers[index] = new Worker(repository.getConnection(), index);
					counts[index] = 0;
				}
			}
		}
	}
	
	private class Worker extends Thread {
		private RepositoryConnection connection;
		private int id;
		
		public Worker(RepositoryConnection connection, int id) {
			this.connection = connection;
			this.id = id;
		}
		
		public void run() {
			TupleQueryResult result = this.connection.prepareTupleQuery(QUERY_FOR_MULTITHREADS[id]).evaluate();
			
			if(result.hasNext()) {
				BindingSet binding = result.next();
				counts[id] = ((Literal)binding.getValue("count")).intValue();
			}
			else
				System.out.println("[Thread" + id + "]Cannot count triangles.");
		}
	}

	@Override
	protected void process() {
		if(!isDirected) {
			connection.prepareUpdate(QUERY_STORE_INITIAL_GRAPH).execute();
			TupleQueryResult result = connection.prepareTupleQuery(QUERY_FIND_TRIANGLES).evaluate();
			
			if(result.hasNext()) {
				BindingSet binding = result.next();
				int count = ((Literal)binding.getValue("count")).intValue();
				writer.writeLine("The number of triangles is " + count);
			}
			else
				System.out.println("Cannot count triangles.");
		}
		else {
			if(!isMultithreaded) {
				TupleQueryResult result = connection.prepareTupleQuery(QUERY_FIND_TRIANGLES_FOR_DIRECTED_GRAPH).evaluate();
				
				if(result.hasNext()) {
					BindingSet binding = result.next();
					int count = ((Literal)binding.getValue("count")).intValue();
					writer.writeLine("The number of triangles is " + count);
				}
				else
					System.out.println("Cannot count triangles.");
			}
			else {
				int count = 0;
				for(int index = 0;index < workers.length;index++) {
					workers[index].start();
				}
				for(int index = 0;index < workers.length;index++) {
					try {
						workers[index].join();
						count += counts[index];
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}
				writer.writeLine("The number of triangles is " + count);
			}
		}
	}

	@Override
	protected void end() {
		if(!this.isDirected)
			connection.prepareUpdate(QUERY_DROP_GRAPH).execute();
	}
}
